### PR TITLE
Zstandard (or any libarchive format and filter) support

### PR DESCRIPTION
**Pull request by mingwandroid**
For discussion.

### CODE REVIEW

**Comment by jcrist** [comment]
> Need an update to the conda.recipe and testing/setup_envs.sh folders as well!

You just need to update the `.travis.yml` file if you want to test with zstd as well. The conda.recipe was old (main recipe is on conda-forge) - just removed it here: https://github.com/conda/conda-pack/commit/6f809009c2440c82d0d3f4ff97adbb6658bf8cee.

**Comment by jcrist** [comment]
A few general comments, questions:

- Is this needed? I understand that `zstd` is supposed to offer faster compression/decompression times, but is the benefit worth the maintenance cost? I suspect the more commonly found `.tar.gz`, `.tar.bz2`, and `.zip` formats will get the bulk of the use, as utilities for extracting these are commonly found on most machines.

- If this work is needed, can this be made an optional dependency? I'd like to keep conda-pack zero dependency if possible, and so far that's been achievable.

**Comment by mingwandroid** [comment]
> Is this needed? I understand that zstd is supposed to offer faster compression/decompression times, but is the benefit worth the maintenance cost? I suspect the more commonly found .tar.gz, .tar.bz2, and .zip formats will get the bulk of the use, as utilities for extracting these are commonly found on most machines.

Well, using libarchive opens you up to a huge array of formats (tar.xz, 7z, 7z sfx etc), yes, tools need to catch up, but that will happen. As for zstandard's performance (it's a lot slower to compress but I didn't take timings for that):

Size:

```
(conda-pack) 0313-rdonnelly:conda-pack rdonnelly$ ls -l devenv.tar.bz2
-rw-------  1 rdonnelly  staff  932510965 30 Sep 03:15 devenv.tar.bz2
(conda-pack) 0313-rdonnelly:conda-pack rdonnelly$ ls -l devenv.tar.zst
-rw-------  1 rdonnelly  staff  441744267 30 Sep 03:50 devenv.tar.zst
```

Time to decompress:

```
time bsdtar -xf devenv.tar.bz2
real    1m58.106s
user    1m43.556s
sys    0m11.183s
time bsdtar -xf devenv.tar.zst
real    0m32.876s
user    0m7.149s
sys    0m16.811s
```

Regarding the maintenance burden well, that's for you to figure out. I don't mind at all if this is rejected.

**Comment by mingwandroid** [comment]
To get a `bsdtar` that can handle `.tar.zst` you can `conda install libarchive`.

**Comment by mingwandroid** [comment]
> If this work is needed, can this be made an optional dependency? I'd like to keep conda-pack zero dependency if possible, and so far that's been achievable.

For sure.

**Comment by mrocklin** [comment]
I don't have enough practical experience with conda-pack to have a strong understanding here, but I'd love to see ZStandard replace bz2/gzip in most applications over time.  The (de)compression time difference can be significant.  I've definitely seen applications that were bound by bz2 decompression costs.  

For conda-pack in particular though I guess we would want to compare the costs of compressing, moving, and decompressing environments relative to other costs in common workflows (For dask this is probably waiting in queues, starting containers, etc..)  I wouldn't be surprised to learn that there were situations where we were bound by moving around and (de)compressing environments, but I also wouldn't be surprised to learn that this was uncommon.

**Comment by mingwandroid** [comment]
I've made zstandard optional now.

**Comment by mingwandroid** [comment]
What's the thinking on this PR @jcrist? If you merge it, I'll follow up with a PR that adds way to use any compression scheme that `libarchive` supports, otherwise I'll close it. Mostly I worked on this to stretch the legs of `libarchive`, though I do think it's a useful and fairly clean addition which opens up the possibility of further cleanup (i.e. there's no need to use either tarfile or zipfile).

**Comment by mingwandroid** [comment]
> For conda-pack in particular though I guess we would want to compare the costs of compressing, moving, and decompressing environments

Yeah, I don't know the scope, but in general I assume things will be uncompressed a lot more than they'll be compressed. I guess for dask those figures are known entities and so an appropriate compression scheme (or level) could be selected dynamically. If you're going to farm the env out to only 2 computers there's not much point spending 15 minutes compressing it, whereas if you're going to farm out to 100 computers then there's probably a big saving in terms of overall energy and bandwidth usage from compressing at the maximum level.

**Comment by mingwandroid** [comment]
List of formats libarchive allows:

```
00 = {str} '7zip'
01 = {str} 'ar_bsd'
02 = {str} 'ar_svr4'
03 = {str} 'cpio'
04 = {str} 'cpio_newc'
05 = {str} 'iso9660'
06 = {str} 'mtree'
07 = {str} 'mtree_classic'
08 = {str} 'pax'
09 = {str} 'pax_restricted'
10 = {str} 'raw'
11 = {str} 'shar'
12 = {str} 'shar_dump'
13 = {str} 'ustar'
14 = {str} 'v7tar'
15 = {str} 'gnutar'
16 = {str} 'warc'
17 = {str} 'xar'
18 = {str} 'zip'
```

List of filters (for tar archives) it supports:

```
00 = {str} 'b64encode'
01 = {str} 'bzip2'
02 = {str} 'compress'
03 = {str} 'grzip'
04 = {str} 'gzip'
05 = {str} 'lrzip'
06 = {str} 'lz4'
07 = {str} 'lzop'
08 = {str} 'none'
09 = {str} 'uuencode'
10 = {str} 'lzip'
11 = {str} 'lzma'
12 = {str} 'xz'
13 = {str} 'zstd'
```

**Comment by msarahan** [comment]
I think adding this to let people play would be a good addition.  We probably shouldn't declare any new format standard yet, but having the capability to support many formats seems very powerful.

**Comment by mingwandroid** [comment]
Well conda-pack doesn't deal with the conda package format at all, just makes an archive in a range of formats. Before this PR, it already supported both zip and tar.bz2.

**Comment by mingwandroid** [comment]
Here is a script that runs various conda-pack with compression schemes and outputs some details:

```
#!/usr/bin/env bash

echo "###########"
echo "# Install #"
echo "###########"

conda deactivate ; pushd /tmp ; rm -rf /tmp/mc3; curl -SLO https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh ; bash ./Miniconda3-latest-MacOSX-x86_64.sh -bfp /tmp/mc3; . /tmp/mc3/bin/activate
pushd ~/conda/conda-pack
  python setup.py install
  conda install -y python-libarchive-c
popd
conda create -y -n python-numpy-scipy python=3.7 numpy scipy

echo "####################"
echo "# Compression time #"
echo "####################"

# tar.bz2 via tarfile
conda-pack -n python-numpy-scipy --output=python-numpy-scipy-tarfile.tar.bz2 --force
# tar.bz2 via libarchive
conda-pack -n python-numpy-scipy --expert-mode --libarchive-format gnutar --libarchive-filter bzip2 --libarchive-options bzip2:compression-level=9 --force
# tar.zst
conda-pack -n python-numpy-scipy --expert-mode --libarchive-format gnutar --libarchive-filter zstd --libarchive-options zstd:compression-level=22 --force
# tar.xz
conda-pack -n python-numpy-scipy --expert-mode --libarchive-format gnutar --libarchive-filter xz --libarchive-options xz:compression-level=9,threads=8 --force
# 7z
conda-pack -n python-numpy-scipy --expert-mode --libarchive-format 7zip --libarchive-options 7zip:compression=lzma2,compression-level=9 --force

echo "############"
echo "# Filesize #"
echo "############"

ls -l python-numpy-scipy*

echo "######################"
echo "# Decompression time #"
echo "######################"

for FMT in -tarfile.tar.bz2 .tar.bz2 .tar.zst .tar.xz .7z; do
   [[ -d /tmp/unpack ]] && rm -rf /tmp/unpack
   mkdir /tmp/unpack
   echo -e "\n\ntime bsdtar -k -C /tmp/unpack -xf python-numpy-scipy$FMT";   time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy$FMT;
done
```

Output:

```
####################
# Compression time #
####################
Collecting packages...
Packing environment at '/tmp/mc3/envs/python-numpy-scipy' to 'python-numpy-scipy-tarfile.tar.bz2'
[########################################] | 100% Completed |  1min  9.0s
Collecting packages...
Packing environment at '/tmp/mc3/envs/python-numpy-scipy' to 'python-numpy-scipy.tar.bz2'
[########################################] | 100% Completed |  1min 11.6s
Collecting packages...
Packing environment at '/tmp/mc3/envs/python-numpy-scipy' to 'python-numpy-scipy.tar.zst'
[########################################] | 100% Completed |  6min 48.5s
Collecting packages...
Packing environment at '/tmp/mc3/envs/python-numpy-scipy' to 'python-numpy-scipy.tar.xz'
[########################################] | 100% Completed |  1min  1.5s
Collecting packages...
Packing environment at '/tmp/mc3/envs/python-numpy-scipy' to 'python-numpy-scipy.7z'
[########################################] | 100% Completed |  4min 28.1s
############
# Filesize #
############
-rw-------  1 rdonnelly  staff  218127642  6 Oct 04:28 python-numpy-scipy-tarfile.tar.bz2
-rw-------  1 rdonnelly  staff  150069354  6 Oct 04:42 python-numpy-scipy.7z
-rw-------  1 rdonnelly  staff  211449181  6 Oct 04:29 python-numpy-scipy.tar.bz2
-rw-------  1 rdonnelly  staff  152139072  6 Oct 04:37 python-numpy-scipy.tar.xz
-rw-------  1 rdonnelly  staff  146468773  6 Oct 04:36 python-numpy-scipy.tar.zst
######################
# Decompression time #
######################


time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy-tarfile.tar.bz2

real	0m25.819s
user	0m23.286s
sys	0m2.241s


time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy.tar.bz2

real	0m28.332s
user	0m25.818s
sys	0m2.204s


time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy.tar.zst

real	0m5.839s
user	0m1.796s
sys	0m2.024s


time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy.tar.xz

real	0m12.914s
user	0m10.577s
sys	0m2.128s


time bsdtar -k -C /tmp/unpack -xf python-numpy-scipy.7z

real	0m12.813s
user	0m10.584s
sys	0m2.091s
```

**Comment by mingwandroid** [comment]
I need to fix the tests here, but I'm not sure how to run them, any hints @jcrist?

**Comment by mcg1969** [comment]
@mingwandroid I personally vote +1 on making this an optional dependency for now. I think we're seeing multiple use cases of conda-pack emerge, each of which might have different interests in compressed size and compression/decompression time. For instance, if we use conda-pack to build an installer, we might favor high compression ratios first, then decompression time. But for cluster distribution we might favor total compression+decompression time.

**Comment by mcg1969** [comment]
@mingwandroid I personally vote +1 on making this an optional dependency for now. I think we're seeing multiple use cases of conda-pack emerge, each of which might have different interests in compressed size and compression/decompression time. For instance, if we use conda-pack to build an installer, we might favor high compression ratios first, then decompression time. But for cluster distribution we might favor total compression+decompression time.

**Comment by jcrist** [comment]
> I need to fix the tests here, but I'm not sure how to run them, any hints @jcrist?

The [travis file](https://github.com/conda/conda-pack/blob/master/.travis.yml) may be helpful. Basically need to build the test environments with `testing/setup_envs.sh`, then it's just `py.test conda-pack`.

**Comment by jcrist** [comment]
Apologies for the delay in review here, it's been a busy couple of weeks.

---

I'm not sold on the new "expert" api being a worthwhile addition - many newer formats are universally better than older formats. I see a usecase for common ones people use for distribution (`.tar`, `.tar.gz`, `.tar.bz2`) as well as newer faster compression schemas (`.tar.zst`) - I'm not sure that others (e.g. `uuencode`) are things we should support. Part of building tools is figuring out sane defaults and being discriminating with options to keep the user api simple.

As such, I think at this time I'd prefer to only add z-standard support with libarchive and keep the existing api the same. I agree that for some use cases the decreased decompression times would be nice. For my use case `tar.gz` is sufficiently performant - I never use the `tar.bz2` version.

**Comment by mcg1969** [comment]
I think that's a reasonable approach. I'm also happy to work on getting this working against master and fixing/updating the tests, e.g., adding `.tar.zst`.

**Comment by mingwandroid** [comment]
Please do @mcg1969, I'm busy with a Qt/OpenSSL 1.1.1 update that we need to unblock openssl 1.1.1.

**Comment by mingwandroid** [comment]
We should think about the compression level setting, bzip uses 0-9 I think while zstd uses 0-22, should we remap a number between 0 and 100 to keep a consistent interface?

**Comment by mcg1969** [comment]
For now, I've dropped the commit that adds `--expert-mode` (but I saved it off to another branch for future consideration).

I did notice that temporary files are being employed for `add_bytes`.  I think it's worth experimenting with getting that to write to the archive directly, as we do for `tarfile`. Ray pointed out that's available in libarchive as `add_file_from_memory`.

**Comment by mcg1969** [comment]
@mingwandroid looks like everything is passing here but Windows Python 2.7, with this error:
```
ArchiveError: Can't launch external program: zstd -22 (errno=22, retcode=-30, archive_p=51097872)
```
Is there another dependency we need to install in the test env?

**Comment by mingwandroid** [comment]
Ah, maybe we don't have zstd support for python 2.7, I thought we did, investigating now.

**Comment by mingwandroid** [comment]
OK I hit a small roadblock, too many tasks in progress at once (5!) so I'm going to leave this until later, I'm only quad-core.

**Comment by wyphan** [comment]
In the meantime, is it possible for `conda-pack` to use `pigz` for multithreaded compression and decompression with the `gz` format?

**Comment by dholth** [comment]
zstandard is amazing. `zstd -T0` for example uses all threads, you will find it is much faster to compress and decompress than gzip with the correct options. https://anaconda.org/conda-forge/zstandard is the best way to access it from Python, you can ditch libarchive and just use the Python standard library. (libarchive does not expose all zstandard options).

I've used conda-pack but lately I do a multistage docker build and --copy-from (build docker image) to (identical path on deployment docker image), avoiding path rewrite bugs

**Comment by dholth** [comment]
I don't think this will be merged as is. Here's how we would do tar.zst with the zstandard library:

```
compres = zstandard.ZstdCompressor(
level=ZSTD_COMPRESS_LEVEL,
threads=ZSTD_COMPRESS_THREADS,
)
with open("somefile.tar.zst", "w") as component_file:
    component_stream = compress.stream_writer(component_file, closefd=False)
    component_tar = tarfile.TarFile(fileobj=component_stream, mode="w")
```

**Comment by dholth** [comment]
I'm not familiar with this codebase. It looks like it has a complex "parallelize compression" system for archives that could al be eliminated with zstandard's builtin compression. This PR does seem to do a lot more than simply "add libarchive + zstandard", we could update the PR to use the binding instead of libarchive and possibly add a benchmark.

File: conda_pack/formats.py [hunk 1]
```diff
@@ -13,244 +13,35 @@
...
-        pass
-
-    def _write_footer(self):
-        pass
-
-    def _flush_compressor(self, compressor):
-        return compressor.flush()
-
-
-class ParallelXZFileWriter(ParallelFileWriter):
```
Comment (dholth) [inline conda_pack/formats.py:217]: these will need to be preserved; or we could simplify and make you use zstd if you need speed

**Comment by dholth** [comment]
I've tried to update this PR. It looks like a different parallel compression support was added between the old PR and today. The old PR does contain some useful generalization. Let me know if you would like to try to continue this PR, but with `python-zstandard` instead of `libarchive`

**Comment by dholth** [comment]
Today you write `zstandard.ZstdCompressor(threads=10)` and you're done, in 2018 they probably had different concerns. I've seen excellent performance from Python's tarfile but haven't compared it with `tar` head to head.

It would probably simplify things to remove the old code, but I don't have a personal opinion about conda-pack.

**Comment by dbast** [comment]
@iamthebot This PR shows how to implement a new format with the existing architecture https://github.com/conda/conda-pack/pull/179 .. should be possible to implement .tar.zstd support this way.

**Comment by math411** [comment]
Hi, was there a decision on direction for this PR? Happy to help with a submission if need!

**Comment by dholth** [comment]
@math411 https://github.com/conda/conda-pack/pull/179/files but with `import zstandard` would be the way to go.

**Pull request closed without merging**